File: ios/Runner/WebViewChannelHandler.swift
import Flutter
import UIKit
import WebKit

/// Handles `com.memexlab.memex/webview` MethodChannel.
class WebViewChannelHandler: NSObject {

    static func register(with messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(
            name: "com.memexlab.memex/webview",
            binaryMessenger: messenger
        )
        let instance = WebViewChannelHandler()
        channel.setMethodCallHandler(instance.handle)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "disableScrolling":
            disableWebViewScrolling()
            result(nil)
        case "renderHtmlToImage":
            renderHtmlToImage(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Render HTML to image (off-screen)

    private func renderHtmlToImage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let html = args["html"] as? String else {
            result(FlutterError(
                code: "INVALID_ARGS",
                message: "renderHtmlToImage requires a 'html' string argument.",
                details: nil
            ))
            return
        }
        let width = (args["width"] as? NSNumber)?.doubleValue ?? 390
        let maxHeight = (args["maxHeight"] as? NSNumber)?.doubleValue ?? 3000

        // The renderer owns a self-contained off-screen WKWebView and keeps a
        // strong reference to itself until the snapshot completes. It is fully
        // isolated from any timeline card WebView, so there is no ambiguity
        // about which WebView is captured.
        let renderer = OffscreenHtmlRenderer(
            width: CGFloat(width),
            maxHeight: CGFloat(maxHeight)
        )
        renderer.render(html: html) { base64 in
            if let base64 = base64 {
                result(base64)
            } else {
                result(FlutterError(
                    code: "RENDER_FAILED",
                    message: "Failed to render HTML snapshot.",
                    details: nil
                ))
            }
        }
    }

    // MARK: - Disable scrolling

    private func disableWebViewScrolling() {
        guard let window = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .flatMap({ $0.windows })
            .first(where: { $0.isKeyWindow }) else { return }
        disableScrollingInView(window)
    }

    private func disableScrollingInView(_ view: UIView) {
        if let webView = view as? WKWebView {
            webView.scrollView.isScrollEnabled = false
            webView.scrollView.showsVerticalScrollIndicator = false
            webView.scrollView.showsHorizontalScrollIndicator = false
            webView.scrollView.bounces = false
            webView.scrollView.alwaysBounceVertical = false
            webView.scrollView.alwaysBounceHorizontal = false
        }
        for subview in view.subviews {
            disableScrollingInView(subview)
        }
    }
}

/// Renders a self-contained HTML document into a PNG (base64) using an
/// off-screen `WKWebView`. The web engine performs the rendering, so the result
/// is reliable on both real devices and simulators (unlike Flutter-layer
/// screenshots of platform views).
private final class OffscreenHtmlRenderer: NSObject, WKNavigationDelegate {

    /// Keeps renderers alive for the duration of the async render.
    private static var active: [OffscreenHtmlRenderer] = []

    private let width: CGFloat
    private let maxHeight: CGFloat
    private var webView: WKWebView?
    private var completion: ((String?) -> Void)?
    private var finished = false

    init(width: CGFloat, maxHeight: CGFloat) {
        self.width = width
        self.maxHeight = maxHeight
        super.init()
    }

    func render(html: String, completion: @escaping (String?) -> Void) {
        self.completion = completion
        OffscreenHtmlRenderer.active.append(self)

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            let config = WKWebViewConfiguration()
            let webView = WKWebView(
                frame: CGRect(x: 0, y: 0, width: self.width, height: 1),
                configuration: config
            )
            webView.isOpaque = false
            webView.backgroundColor = .clear
            webView.scrollView.backgroundColor = .clear
            webView.scrollView.isScrollEnabled = false
            webView.navigationDelegate = self
            self.webView = webView

            // Insert behind everything so it lays out/renders without being
            // visible to the user, then remove it once captured.
            if let window = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .flatMap({ $0.windows })
                .first(where: { $0.isKeyWindow }) {
                window.insertSubview(webView, at: 0)
            }

            webView.loadHTMLString(html, baseURL: nil)

            // Safety timeout in case navigation never finishes.
            DispatchQueue.main.asyncAfter(deadline: .now() + 8) { [weak self] in
                self?.complete(with: nil)
            }
        }
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        // Measure real content height, then resize and snapshot.
        webView.evaluateJavaScript("document.body.scrollHeight") { [weak self] value, _ in
            guard let self = self else { return }
            var height = self.maxHeight
            if let n = value as? NSNumber {
                height = min(CGFloat(n.doubleValue), self.maxHeight)
            }
            if height <= 0 { height = 1 }
            webView.frame = CGRect(x: 0, y: 0, width: self.width, height: height)

            // Let layout settle (images/fonts) before capturing.
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.15) { [weak self] in
                self?.takeSnapshot()
            }
        }
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        complete(with: nil)
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        complete(with: nil)
    }

    private func takeSnapshot() {
        guard let webView = webView else {
            complete(with: nil)
            return
        }
        let config = WKSnapshotConfiguration()
        config.rect = CGRect(origin: .zero, size: webView.frame.size)
        webView.takeSnapshot(with: config) { [weak self] image, _ in
            guard let self = self else { return }
            guard let image = image, let data = image.pngData() else {
                self.complete(with: nil)
                return
            }
            self.complete(with: data.base64EncodedString())
        }
    }

    private func complete(with base64: String?) {
        if finished { return }
        finished = true
        webView?.navigationDelegate = nil
        webView?.removeFromSuperview()
        webView = nil
        let cb = completion
        completion = nil
        cb?(base64)
        OffscreenHtmlRenderer.active.removeAll { $0 === self }
    }
}

File: android/app/src/main/kotlin/com/memexlab/memex/channels/WebViewChannelHandler.kt
package com.memexlab.memex.channels

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.webkit.WebViewClient
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream
import kotlin.math.min

/**
 * Handles `com.memexlab.memex/webview` MethodChannel.
 */
class WebViewChannelHandler(private val activity: Activity) {

    companion object {
        private const val CHANNEL = "com.memexlab.memex/webview"

        fun register(flutterEngine: FlutterEngine, activity: Activity) {
            val handler = WebViewChannelHandler(activity)
            MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL)
                .setMethodCallHandler { call, result ->
                    when (call.method) {
                        "disableScrolling" -> {
                            handler.disableWebViewScrolling()
                            result.success(null)
                        }
                        "renderHtmlToImage" -> {
                            val html = call.argument<String>("html")
                            if (html == null) {
                                result.error(
                                    "INVALID_ARGS",
                                    "renderHtmlToImage requires a 'html' string argument.",
                                    null
                                )
                            } else {
                                val width = call.argument<Double>("width") ?: 390.0
                                val maxHeight = call.argument<Double>("maxHeight") ?: 3000.0
                                handler.renderHtmlToImage(html, width, maxHeight, result)
                            }
                        }
                        else -> result.notImplemented()
                    }
                }
        }
    }

    /**
     * Renders a self-contained HTML document into a PNG (base64) using an
     * off-screen [WebView]. Rendering happens inside the native WebView, so
     * `webView.draw(canvas)` captures real content (unlike Flutter-layer
     * screenshots of platform views, which can be blank under hybrid
     * composition). The WebView is created/owned here and never attached to the
     * visible view tree, so it is fully isolated from timeline card WebViews.
     */
    fun renderHtmlToImage(
        html: String,
        width: Double,
        maxHeight: Double,
        result: MethodChannel.Result,
    ) {
        val mainHandler = Handler(Looper.getMainLooper())
        mainHandler.post {
            val density = activity.resources.displayMetrics.density
            val widthPx = (width * density).toInt().coerceAtLeast(1)
            val maxHeightPx = (maxHeight * density).toInt().coerceAtLeast(1)

            val webView = WebView(activity)
            // Single-shot guard so we resolve the Flutter result exactly once.
            var settled = false
            fun finish(base64: String?, errorMessage: String?) {
                if (settled) return
                settled = true
                try {
                    webView.destroy()
                } catch (_: Throwable) {
                }
                if (base64 != null) {
                    result.success(base64)
                } else {
                    result.error(
                        "RENDER_FAILED",
                        errorMessage ?: "Failed to render HTML snapshot.",
                        null
                    )
                }
            }

            try {
                webView.settings.javaScriptEnabled = true
                webView.settings.loadWithOverviewMode = false
                webView.settings.useWideViewPort = false
                webView.setBackgroundColor(Color.TRANSPARENT)

                webView.webViewClient = object : WebViewClient() {
                    override fun onPageFinished(view: WebView, url: String) {
                        // Measure real content height, then layout + capture.
                        view.evaluateJavascript("document.body.scrollHeight") { value ->
                            val cssHeight = value?.toFloatOrNull() ?: maxHeight.toFloat()
                            val heightPx = min((cssHeight * density).toInt(), maxHeightPx)
                                .coerceAtLeast(1)

                            view.measure(
                                View.MeasureSpec.makeMeasureSpec(widthPx, View.MeasureSpec.EXACTLY),
                                View.MeasureSpec.makeMeasureSpec(heightPx, View.MeasureSpec.EXACTLY),
                            )
                            view.layout(0, 0, widthPx, heightPx)

                            // Let layout/images settle before drawing.
                            mainHandler.postDelayed({
                                try {
                                    val bitmap = Bitmap.createBitmap(
                                        widthPx,
                                        heightPx,
                                        Bitmap.Config.ARGB_8888,
                                    )
                                    val canvas = Canvas(bitmap)
                                    view.draw(canvas)
                                    val stream = ByteArrayOutputStream()
                                    bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
                                    bitmap.recycle()
                                    val base64 = Base64.encodeToString(
                                        stream.toByteArray(),
                                        Base64.NO_WRAP,
                                    )
                                    finish(base64, null)
                                } catch (e: Throwable) {
                                    finish(null, e.message)
                                }
                            }, 150)
                        }
                    }
                }

                webView.loadDataWithBaseURL(null, html, "text/html", "UTF-8", null)

                // Safety timeout in case onPageFinished never fires.
                mainHandler.postDelayed({
                    finish(null, "Render timed out.")
                }, 8000)
            } catch (e: Throwable) {
                finish(null, e.message)
            }
        }
    }

    private fun disableWebViewScrolling() {
        val rootView = activity.window.decorView.rootView
        disableScrollingInView(rootView)
    }

    private fun disableScrollingInView(view: View) {
        if (view is WebView) {
            view.isVerticalScrollBarEnabled = false
            view.isHorizontalScrollBarEnabled = false
            view.overScrollMode = View.OVER_SCROLL_NEVER
        }
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                disableScrollingInView(view.getChildAt(i))
            }
        }
    }
}
